content/blog/2019-09-25-vril-vector-bril.md

```diff
@@ -0,0 +1,139 @@
+
+### Initializing arrays in Vril
+Vril extends the grammar of Bril to include arrays by adding an `init` operator to initialize an array of length `l` as follows,
+```
+array_name : array = init l
+```
+
+Vril adds array operations to Bril of the form,
+```
+array/ variable: type = aop args
+```
+### Supported array operations
+Vril can operate on arrays in two ways: 1) We can perform *scalar* accesses to arrays by moving one element of the array to a variable and performing any arithmetic or logic operation that are already supported in the original bril, and 2) we can perform *vector* accesses to the arrays by using new vector operations that take arrays as arguments directly.
+
+In order to move data from an element of an array `arr[idx]` to a variable `var`, we can use `a2v` op:
+```
+var: int = a2v arr idx
+```
+Similarly, we can move a value stored in a variable `var` into one of the array's element `arr[idx]`:
+```
+arr: int v2a var index
+```
+
+To put a constant into an array, the constant should first be put to a variable.
+To operate with a scalar operator (for instance `print`) on a single element, the array element has to first be put to a variable.
+
+Vector ops
+----------------------------
+
+### Vector ops in Vril
+There are two types of vector ops in Vril: configuration ops and arithmetic/logic ops. Vector arithmetic or logic operations take arrays as arguments and produce an array or a scalar value as a result. Instead, configuration ops help in configuring a vector state.
+
+### Adding vector ops to Vril
+For now there is only one configuration op: `setvl`. 
+```
+vl: int setvl val
+```
+is used to request `val` lanes to the IR and `setvl` returns `vl:=val` if `vl<maxvl` or `vl:=maxvl` if `vl>=maxvl`.
+We have also implemented only one arithmetic vector op `vadd` as a proof of concept:
+```
+arr3: array vadd arr1 arr2
+```
+Which takes two arrays as arguments and performs element-wise vector addition, and stores it into a third array.
+
+### Simulating vector ops in Vrili (Vril interpreter)
+We have extended Brili to interpret vector operations.
+Vrili executes for loops across the different elements of the array operands in order to produce the correct result. This emulates what a vector microarchitecture would do concurrently with a single fetch-decode effort.
+
+Evaluation
+----------------------------
+
+Vril makes two major additions to Bril, adding Arrays and extending Briloperations with vector operations. This section describes how we evaluate each of these.
+
+### Evaluating Arrays
+We take three approaches to evaluate arrays,
+- evaluating integration to Bril
+- qualitative evaluation to justify arrays
+- evaluating functionality of arrays
+
+#### Integration tests for arrays
+We ran tests to generate JSON representation of Bril from text format and vice-versa to evaluate whether the array specification is expressible in both available formats. These tests reside in `vtest/parse` and `vtest/print` directories.
```

Comment:
To be clear, this _just_ tests the parser and pretty-printer. So I don't think I would call this "evaluating integration to Bril"; it's just testing the text representation machinery alone.

Reply:
My expectation is to gain confidence that our changes to Bril IR reflect our intentions. I'm trying to do that by writing a program and inspecting the generated IR from the parser.